source/dotnet/Library/AdaptiveCards/AdaptiveTypedElementConverter.cs

```diff
@@ -66,84 +55,115 @@ public static void RegisterTypedElement<T>(string typeName = null)
             TypedElementTypes.Value[typeName] = typeof(T);
         }
 
-        public override bool CanRead => true;
-
-        public override bool CanWrite => false;
-
         public override bool CanConvert(Type objectType)
         {
             return typeof(AdaptiveTypedElement).GetTypeInfo().IsAssignableFrom(objectType.GetTypeInfo());
         }
 
+        public override bool CanWrite => false;
         public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
         {
             throw new NotImplementedException();
         }
 
+        public override bool CanRead => true;
         public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
         {
             var jObject = JObject.Load(reader);
 
-            var typeName = jObject["type"]?.Value<string>() ?? jObject["@type"]?.Value<string>();
-            if (typeName == null)
-            {
-                // Get value of this objectType's "Type" JsonProperty(Required)
-                var typeJsonPropertyRequiredValue = objectType.GetRuntimeProperty("Type")
-                    .CustomAttributes.Where(a => a.AttributeType == typeof(JsonPropertyAttribute)).FirstOrDefault()?
-                    .NamedArguments.Where(a => a.TypedValue.ArgumentType == typeof(Required)).FirstOrDefault()
-                    .TypedValue.Value.ToString();
-
-                // If this objectType does not require "Type" attribute, use the objectType's XML "TypeName" attribute
-                if (typeJsonPropertyRequiredValue == "0")
-                {
-                    typeName = objectType
-                        .GetRuntimeFields().Where(x => x.Name == "TypeName").FirstOrDefault()?
-                        .GetValue("TypeName").ToString();
-                }
-                else
-                {
-                    throw new AdaptiveSerializationException("Required property 'type' not found on adaptive card element");
-                }
-            }
+            string typeName = GetElementTypeName(objectType, jObject);
 
             if (TypedElementTypes.Value.TryGetValue(typeName, out var type))
             {
-                if (jObject.Value<string>("id") == null)
+                string objectId = jObject.Value<string>("id");
+                if (objectId == null)
```

Comment:
>if (objectId == null) [](start = 16, length = 21)

do we want this check to be include empty/whitespace values for id? so objectId.isNullOrWhitespace() ?

Reply:
hmm... I don't think we currently ban the use of whitespace for ids. we probably *should* ban it. I filed #2748 to investigate.

---
In reply to: [278323340](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278323340) [](ancestors = 278323340)